File: 01/ruby/part02.rb
# Day 1: The Tyranny of the Rocket Equation
# https://adventofcode.com/2019/day/1

# load the input
input = File.open("./inputs/01.txt")

# define a method to calculate the fuel based on mass
def calculate_fuel(mass)
  ((mass.to_i / 3) - 2)
end

# iterate over each line of input
total = input.each_line.reduce(0) do |fuel, module_mass|
  # calculate the module fuel
  module_fuel     = calculate_fuel(module_mass)
  # calculate the first additional fuel
  additional_fuel = calculate_fuel(module_fuel)

  # whilst we have positive additional fuel
  while additional_fuel > 0
    # add it to the modules total
    module_fuel += additional_fuel
    # recalculate the additional fuel on the latest additional fuel
    additional_fuel = calculate_fuel(additional_fuel)
  end

  # add the module fuel to the previous fuel value
  fuel + module_fuel
end

# output the fuel
puts total

File: 01/go/part02.go
package main

import (
	"bufio"
	"fmt"
	"os"
	"strconv"
)

func fuelRequirement(mass int) int {
	if mass := (mass / 3) - 2; mass > 0 {
		return mass + fuelRequirement(mass)
	}
	return 0
}

func main() {
	f, _ := os.Open("./inputs/01.txt")
	defer f.Close()

	var (
		total      int
		moduleMass int
	)

	rdr := bufio.NewReader(f)
	for {
		line, err := rdr.ReadString('\n')
		if err != nil {
			break
		}
		moduleMass, _ = strconv.Atoi(line[:len(line)-1])
		total += fuelRequirement(moduleMass)
	}

	fmt.Println(total)
}
